transport/src/main/java/io/netty/channel/AbstractFileRegion.java

```diff
@@ -0,0 +1,184 @@
+                    + " (expected: 0 - " + (this.count() - 1) + ')');
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("negative length " + length);
+        }
+        length = Math.min(count() - position, length);
+        if (length == 0) {
+            return 0L;
+        }
+        return channel().transferTo(position() + position, length, target);
+    }
+
+    /**
+     * Returns a slice of this file region's sub-region. This method does not modify
+     * {@code transferIndex} of this buffer.
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the reference count
+     * will NOT be increased.
+     */
+    public AbstractFileRegion slice(long index, long length) {
+        return new SlicedFileRegion(this, index, length);
+    }
+
+    /**
+     * Returns a slice of this file region's sub-region. This method does not modify
+     * {@code transferIndex} of this buffer.
+     * <p>
+     * Note that this method returns a {@linkplain #retain() retained} file region unlike
+     * {@link #slice()}.
+     */
+    public AbstractFileRegion retainedSlice(long index, long length) {
+        AbstractFileRegion sliced = slice(index, length);
+        sliced.retain();
+        return sliced;
+    }
+
+    /**
+     * Returns a new slice of this file region's sub-region starting at the current
+     * {@code transferIndex} and increases the {@code transferIndex} by the size of the new slice (=
+     * {@code length}).
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the reference count
+     * will NOT be increased.
+     *
+     * @param length
+     *            the size of the new slice
+     */
+    public AbstractFileRegion transferSlice(long length) {
+        AbstractFileRegion sliced = slice(transferIndex(), length);
+        transferIndex += length;
+        return sliced;
+    }
+
+    /**
+     * Returns a new slice of this file region's sub-region starting at the current
+     * {@code transferIndex} and increases the {@code transferIndex} by the size of the new slice (=
+     * {@code length}).
+     * <p>
+     * Note that this method returns a {@linkplain #retain() retained} file region unlike
+     * {@link #transferSlice()}.
+     *
+     * @param length
+     *            the size of the new slice
+     */
+    public AbstractFileRegion transferRetainedSlice(long length) {
+        AbstractFileRegion sliced = transferSlice(length);
+        sliced.retain();
+        return sliced;
+    }
+
+    /**
+     * Return the underlying file region instance if this file region is a wrapper of another file
+     * region.
+     *
+     * @return {@code null} if this file region is not a wrapper
+     */
+    public abstract AbstractFileRegion unwrap();
+
+    /**
+     * Return the underlying file channel instance of this file region.
+     */
+    public abstract FileChannel channel() throws IOException;
```

Comment:
I really dislike that we need to expose this at all as a FileRegion should not really depend on a FileChannel. This is an implementation detail.